examples/src/BiddingGame.hs

```diff
@@ -0,0 +1,175 @@
+--    the money among the winners
+-- 5. Players are able to get their money back if the operator
+--    does not publish the game results before the publishing
+--    deadline.
+--
+-- The operator can only close the game once. In particular, she
+-- shouldn't be able to close the game more than once with
+-- different sets of bids.
+--
+-- Bids should be possible to do in parallel. Note that this
+-- requirement makes impossible:
+-- * to guarantee that the operator doesn't leave any bids unconsumed.
+-- * to stop bidders from bidding more than once.
+--
+-- The operator must collect the bids ahead of knowing the result
+-- of the game. Otherwise, she could chose to only collect the bids of
+-- the losers and never pay any money back to the winners.
+--
+module BiddingGame where
+
+import Data.Aeson (FromJSON, ToJSON)
+import GHC.Generics (Generic)
+import qualified Ledger
+import qualified Ledger.Ada as Ada
+import qualified Ledger.Typed.Scripts as Scripts
+import qualified Ledger.Value as Value (geq)
+import Plutus.V1.Ledger.Api (Datum (Datum))
+import Plutus.V1.Ledger.Contexts (ScriptContext (..))
+import qualified PlutusTx
+import PlutusTx.Prelude hiding (Applicative (..))
+import Schema (ToSchema)
+import qualified Prelude as Haskell
+
+
+-- * On-chain validation
+--
+-- $contract-flow
+-- Contract flow
+--
+-- * The operator creates an output to the script with datum @GameStart@
+-- * The bidders pay to the script with datum @Bid GameResult@
+-- * The operator collects the bids and pays to the script with all of
+--   the collected money and datum @CollectedBids _bids@.
+-- * The operator publishes the result and pays to the winners.
+--
+-- If the operator does not collect, players reclaim their individual bids.
+-- If the operator does not publish the result, anyone can return the
+-- collected bids to all players.
+
+data BidParams = BidParams
+  { biddingDeadline :: Ledger.POSIXTime
+  , publishingDeadline :: Ledger.POSIXTime
+  , minimumBid :: Integer
+  , description :: BuiltinByteString
+  , operator :: Ledger.PubKeyHash
+  }
+  deriving stock (Haskell.Show, Haskell.Eq, Generic)
+  deriving anyclass (ToJSON, FromJSON, ToSchema)
+
+type GameResult = Bool
+data BidDatum
+  = GameStart BidParams
+  | Bid GameResult
+  | -- | Contains the bidders, with duplicates if they bidded more than once.
+    CollectedBids [(Ledger.PubKeyHash, GameResult, Integer)]
+  deriving stock (Haskell.Show, Haskell.Eq, Generic)
+
+data BidRedeemer
+  = BidCollection
+  | GameClose GameResult
+  | BidReclaim
+
+minLovelace :: Integer
+minLovelace = 2000000
+
+validateBid
+  :: BidParams -> BidDatum -> BidRedeemer -> ScriptContext -> Bool
+validateBid p d r ctx =
+    let info :: Ledger.TxInfo
+        info = Ledger.scriptContextTxInfo ctx
+
+        consumesGameStart =
+          isJust $
+            Ledger.findDatumHash
+              (Datum $ PlutusTx.toBuiltinData $ GameStart p)
+              info
```

Comment:
I want `consumesGameStart` to be `True` when there is an input in the pending transaction which consumes an output with datum `GameStart p`.

I thought also of testing `txInfoInputs` input by input and comparing the datum hashes, however, I couldn't find in the API how to compute a DatumHash from a Datum!

We discovered last week that this requirement accomplishes no purpose. It was originally intended to ensure the operator only closes the game once. But still I'm curious of how one would accomplish something like this.

Reply:
I rephrased the code since asking. Perhaps it is cheaper to lookup the hash than to actually compute it (?)